client.go

```diff
@@ -0,0 +1,410 @@
+func NewClientWithConfig(configProvider ConfigProvider) (*Client, error) {
+	if config, err := configProvider.Config(); err != nil {
+		return nil, err
+	} else {
+		return newClient("", *config)
+	}
+}
+
+type Client struct {
+	// configuration
+	name          string
+	clusterConfig *cluster.Config
+
+	// components
+	proxyManager        *proxyManager
+	connectionManager   *icluster.ConnectionManager
+	clusterService      *icluster.ServiceImpl
+	partitionService    *icluster.PartitionService
+	eventDispatcher     *event.DispatchService
+	userEventDispatcher *event.DispatchService
+	invocationHandler   invocation.Handler
+	logger              ilogger.Logger
+
+	// state
+	state    atomic.Value
+	refIDGen *iproxy.ReferenceIDGenerator
+}
+
+func newClient(name string, config Config) (*Client, error) {
+	config = config.clone()
+	id := atomic.AddInt32(&nextId, 1)
+	if name == "" {
+		name = fmt.Sprintf("hz.client_%d", id)
+	}
+	// TODO: consider disabling manual client name
+	if config.ClientName != "" {
+		name = config.ClientName
+	}
+	logLevel, err := ilogger.GetLogLevel(config.LoggerConfig.Level)
+	if err != nil {
+		return nil, err
+	}
+	clientLogger := ilogger.NewWithLevel(logLevel)
+	client := &Client{
+		name:                name,
+		clusterConfig:       &config.ClusterConfig,
+		eventDispatcher:     event.NewDispatchService(),
+		userEventDispatcher: event.NewDispatchService(),
+		logger:              clientLogger,
+		refIDGen:            iproxy.NewReferenceIDGenerator(),
+	}
+	client.state.Store(created)
+	client.subscribeUserEvents()
+	client.createComponents(&config)
+	return client, nil
+}
+
+// Name returns client's name
+// Use ConfigBuilder.SetName to set the client name.
+// If not set manually, an automatic name is used.
+func (c *Client) Name() string {
+	return c.name
+}
+
+// GetMap returns a distributed map instance.
+func (c *Client) GetMap(name string) (*Map, error) {
+	return c.GetMapContext(context.Background(), name)
+}
+
+// GetMapContext returns a distributed map instance.
+func (c *Client) GetMapContext(ctx context.Context, name string) (*Map, error) {
+	if !c.ready() {
+		return nil, ErrClientNotReady
+	}
+	if ctx == nil {
+		return nil, ErrContextIsNil
+	}
+	return c.proxyManager.getMapWithContext(ctx, name)
+}
+
+// GetReplicatedMap returns a replicated map instance.
+func (c *Client) GetReplicatedMap(name string) (*ReplicatedMap, error) {
+	if !c.ready() {
+		return nil, ErrClientNotReady
+	}
+	return c.GetReplicatedMapContext(context.Background(), name)
+}
+
+// GetReplicatedMapContext returns a replicated map instance.
+func (c *Client) GetReplicatedMapContext(ctx context.Context, name string) (*ReplicatedMap, error) {
+	if !c.ready() {
+		return nil, ErrClientNotReady
+	}
+	if ctx == nil {
+		return nil, ErrContextIsNil
+	}
+	return c.proxyManager.getReplicatedMapWithContext(ctx, name)
+}
+
+// Start connects the client to the cluster.
+func (c *Client) Start() error {
+	if !c.canStart() {
```

Comment:
This check against running `client.Start()` in multiple goroutines in parallel. Such code would lead to a race with undefined behavior, so we should protect this method with a mutex. The same consideration applies to `Shutdown()` method.

Reply:
Thanks, I've unexported start and added a race test to shutdown